### ISSUE TITLE
Proper lifting of atomic operations

### ISSUE THREAD

**Issue by alexrp** [Vector35/binaryninja-api#5066]
As I'm delving deeper into multithreaded code, the lack of proper lifting for atomic operations is starting to become a bit of a problem.

Some examples:

![image](https://github.com/Vector35/binaryninja-api/assets/44076/a48e2db9-8fb6-47d4-bb14-147a1244b517)
![image](https://github.com/Vector35/binaryninja-api/assets/44076/e5dd9033-b92c-49b2-a91d-6c0c650d80bd)

There's no obvious indicator that these are actually `lock cmpxchg` and `lock inc` respectively. You'd only realize that if you switch over to disassembly. This can lead to bugs if I start writing code that interoperates with the decompiled code in question but doesn't do operations atomically because I missed that detail.

(Is this already tracked somewhere? I seem to recall it being discussed on other issues, but couldn't find anything.)

**Comment by yrp604** [comment]
It's not an answer but same question here: https://github.com/Vector35/arch-arm64/discussions/117

**Comment by adfernandes** [comment]
Yes, on [Arm-v8**M**](https://developer.arm.com/documentation/100235/0004/the-cortex-m33-instruction-set/memory-access-instructions/ldaex-and-stlex) the atomic operations interact with the "Global Exclusivity Monitor", which exists _outside_ the CPU, and is usually on the main bus fabric crossbar so that writes to memory can be coordinated.

When you lift and optimize this code, unless you realize that there are "side effects", a higher-level optimization pass can remove the operation since it looks like a "read what I just wrote" memory access.
